src/Microsoft.ServiceFabric.Actors/Runtime/MigrationActorStateProvider.cs

```diff
@@ -0,0 +1,445 @@
+        }
+
+        /// <inheritdoc/>
+        public Task RestoreAsync(string backupFolderPath)
+        {
+            return ((IStateProviderReplica)this.rcStateProvider).RestoreAsync(backupFolderPath);
+        }
+
+        /// <inheritdoc/>
+        public Task RestoreAsync(string backupFolderPath, RestorePolicy restorePolicy, CancellationToken cancellationToken)
+        {
+            return ((IStateProviderReplica)this.rcStateProvider).RestoreAsync(backupFolderPath, restorePolicy, cancellationToken);
+        }
+
+        /// <inheritdoc/>
+        public Task OnSnapshotAsync(TimeSpan currentLogicalTime)
+        {
+            return ((VolatileLogicalTimeManager.ISnapshotHandler)this.rcStateProvider).OnSnapshotAsync(currentLogicalTime);
+        }
+
+        /// <summary>
+        /// Modifies data from KVS store into a suitable format for RC. Saves the modified data in RC store.
+        /// </summary>
+        /// <param name="keyValuePairs">
+        /// Data from KVS store that needs to be modified and saved in RC
+        /// </param>
+        /// <param name="cancellationToken">
+        /// Cancellation token
+        /// </param>
+        /// <returns>A <see cref="Task"/> representing the result of the asynchronous operation.</returns>
+        public async Task SaveStatetoRCAsync(List<KeyValuePair<string, byte[]>> keyValuePairs, CancellationToken cancellationToken)
+        {
+            List<RCData> dataToSave = new List<RCData>();
+            foreach (KeyValuePair<string, byte[]> pair in keyValuePairs)
+            {
+                byte[] rcValue = { };
+                IReliableDictionary2<string, byte[]> dictionary = null;
+                if (this.IsPresenceKey(pair.Key))
+                {
+                    rcValue = pair.Value;
+                    var index = pair.Key.LastIndexOf('_');
+                    var actorId = new ActorId(pair.Key.Substring(index + 1));
+                    dictionary = this.rcStateProvider.GetActorStateDictionary(actorId);
+                }
+                else if (this.IsActorStateKey(pair.Key))
+                {
+                    rcValue = pair.Value;
+                    var startIndex = this.GetNthIndex(pair.Key, '_', 2);
+                    var endIndex = this.GetNthIndex(pair.Key, '_', 3);
+                    var actorId = new ActorId(pair.Key.Substring(startIndex + 1, endIndex - startIndex - 1));
+                    dictionary = this.rcStateProvider.GetActorStateDictionary(actorId);
+                }
+                else if (this.IsReminderCompletedKey(pair.Key))
+                {
+                    ReminderCompletedData reminderCompletedData = this.DeserializeReminderCompletedData(pair.Value);
+                    rcValue = this.SerializeReminderCompletedData(reminderCompletedData);
+                    var startIndex = this.GetNthIndex(pair.Key, '_', 2);
+                    var endIndex = this.GetNthIndex(pair.Key, '_', 3);
```

Comment:
Not using this helper since it essentially returns the second part of a string split by underscore. Here we need to evaluate the actorID based on the key type which may not always be the second part of a string split by underscore